File: packs/dsa-trees-ts/challenges/06-binary-tree-paths.test.ts
import { join } from 'node:path';
import { writeChallengeToTmp, importModule } from '../../test-helpers.js';

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function buildTree(values: (number | null)[]): TreeNode | null {
  if (!values.length || values[0] === null) return null;
  const root = new TreeNode(values[0]);
  const queue: TreeNode[] = [root];
  let i = 1;
  while (queue.length && i < values.length) {
    const node = queue.shift()!;
    if (i < values.length && values[i] !== null) {
      node.left = new TreeNode(values[i] as number);
      queue.push(node.left);
    }
    i++;
    if (i < values.length && values[i] !== null) {
      node.right = new TreeNode(values[i] as number);
      queue.push(node.right);
    }
    i++;
  }
  return root;
}

function treeToList(root: TreeNode | null): (number | null)[] {
  if (!root) return [];
  const result: (number | null)[] = [];
  const queue: (TreeNode | null)[] = [root];
  while (queue.length) {
    const node = queue.shift()!;
    if (node) {
      result.push(node.val);
      queue.push(node.left);
      queue.push(node.right);
    } else {
      result.push(null);
    }
  }
  while (result.length && result[result.length - 1] === null) result.pop();
  return result;
}

const CHALLENGE = join(import.meta.dirname, '06-binary-tree-paths.json');

describe('Binary Tree Paths', () => {
  let binaryTreePaths: (root: TreeNode | null) => string[];
  let cleanup: () => void;

  beforeAll(async () => {
    const tmp = writeChallengeToTmp(CHALLENGE);
    cleanup = tmp.cleanup;
    const mod = await importModule<{ binaryTreePaths: typeof binaryTreePaths }>(
      join(tmp.tmpDir, 'solution.ts')
    );
    binaryTreePaths = mod.binaryTreePaths;
  });

  afterAll(() => cleanup());

  it('returns multiple paths', () => {
    expect(binaryTreePaths(buildTree([1, 2, 3, null, 5])).sort()).toEqual(
      ['1->2->5', '1->3'].sort()
    );
  });

  it('returns single path for single node', () => {
    expect(binaryTreePaths(buildTree([1]))).toEqual(['1']);
  });

  it('returns empty array for null root', () => {
    expect(binaryTreePaths(null)).toEqual([]);
  });

  it('returns single path for left-only chain', () => {
    expect(binaryTreePaths(buildTree([1, 2, null, 3]))).toEqual(['1->2->3']);
  });

  it('returns two paths for full tree', () => {
    expect(binaryTreePaths(buildTree([1, 2, 3])).sort()).toEqual(['1->2', '1->3'].sort());
  });
});

File: packs/dsa-trees-py/challenges/06-binary-tree-paths.test.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree(values):
    """Build binary tree from level-order list. None means no node."""
    if not values or values[0] is None:
        return None
    root = TreeNode(values[0])
    queue = deque([root])
    i = 1
    while queue and i < len(values):
        node = queue.popleft()
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    return root


def tree_to_list(root):
    """Convert tree to level-order list (with None for missing nodes)."""
    if not root:
        return []
    result = []
    queue = deque([root])
    while queue:
        node = queue.popleft()
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    while result and result[-1] is None:
        result.pop()
    return result


from solution import binary_tree_paths


def test_multiple_paths():
    root = build_tree([1, 2, 3, None, 5])
    assert sorted(binary_tree_paths(root)) == sorted(["1->2->5", "1->3"])


def test_single_node():
    root = build_tree([1])
    assert binary_tree_paths(root) == ["1"]


def test_empty_tree():
    assert binary_tree_paths(None) == []


def test_left_only_path():
    root = build_tree([1, 2, None, 3])
    assert binary_tree_paths(root) == ["1->2->3"]


def test_full_tree():
    root = build_tree([1, 2, 3])
    assert sorted(binary_tree_paths(root)) == sorted(["1->2", "1->3"])
